src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java

```diff
@@ -304,69 +295,59 @@ public int[] getRequiredTokens() {
     @SuppressWarnings("deprecation")
     @Override
     public void beginTree(DetailAST rootAST) {
-        matcher = format.matcher(getFileContents().getText().getFullText());
-        matchCount = 0;
-        errorCount = 0;
-        findMatch();
+        final Matcher matcher = format.matcher(getFileContents().getText().getFullText());
+        processRegexpMatches(matcher);
     }
 
     /**
-     * Recursive method that finds the matches.
+     * Finds the matches in the file.
      *
-     * @noinspection TailRecursion
-     * @noinspectionreason TailRecursion - until issue #14814
+     * @param matcher The matcher
      */
     // suppress deprecation until https://github.com/checkstyle/checkstyle/issues/11166
+    // -@cs[CyclomaticComplexity] Most simplified logic
     @SuppressWarnings("deprecation")
-    private void findMatch() {
-        final boolean foundMatch = matcher.find();
-        if (foundMatch) {
+    private void processRegexpMatches(Matcher matcher) {
+        // Tracks number of errors.
+        int errorCount = 0;
+        int matchCount = 0;
+        while (errorCount <= errorLimit - 1 && matcher.find()) {
             final FileText text = getFileContents().getText();
```

Comment:
Could you please check if `text` is the same for all iterations. If yes, let's move it outside the `while` loop.

Reply:
yep, you are right . moved it outside the loop